<a href="https://colab.research.google.com/github/Andree-trg/Pemerograman-Dasar/blob/main/Latihan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
Andre_Gerardo_Tarigan = pd.read_csv('/content/mahasiswa.csv')
#import file csv

In [ ]:
Andre_Gerardo_Tarigan.head()
#melihat 5 data teratas

nim   nama           jurusan    umur nilai_akhir  kehadiran_pct  \
0  E0001   Aldi            Teknik     150        59.5           82.0   
1  E0002   Budi            teknik    20.0        84.5           99.0   
2  E0003  Citra            Energi      -5        56.2           78.0   
3  E0004   Dewi            Teknik  twenty        63.0           99.0   
4  E0005    Eka  Manajemen Energi    22.0         999           65.0   

  tanggal_daftar  
0     2024-05-05  
1     2024-02-14  
2     2024-04-29  
3     2024-08-02  
4     2024-03-22

Pada tahap awal analisis data, beberapa library penting terlebih dahulu di-import, yaitu numpy, pandas, matplotlib.pyplot, dan seaborn, yang masing-masing berguna untuk pengolahan numerik, manipulasi tabel data, serta pembuatan visualisasi. Setelah itu, dataset mahasiswa.csv dibaca menggunakan fungsi pd.read_csv() dan dimuat ke dalam sebuah DataFrame bernama Andre_Gerardo_Tarigan. Untuk memperoleh gambaran awal mengenai kondisi data, digunakan fungsi .head() yang menampilkan lima baris teratas dari dataset. Dari tampilan tersebut, dapat dilihat bahwa data masih mengandung berbagai ketidaksesuaian dan error yang menunjukkan bahwa dataset ini belum siap untuk dianalisis lebih jauh dan harus melalui proses pembersihan. Misalnya, pada kolom jurusan terlihat ketidakseragaman penulisan seperti “Teknik” dan “teknik” yang seharusnya mewakili kategori yang sama tetapi ditulis dengan format berbeda. Pada kolom umur, ditemukan nilai yang tidak realistis seperti 150 tahun, umur negatif -5, serta nilai non-numerik seperti “twenty”, yang menunjukkan adanya kesalahan input maupun ketidaksesuaian tipe data. Pada kolom nilai_akhir, terdapat nilai ekstrem 999 yang berada jauh di luar rentang nilai normal 0–100, sehingga dapat disebut sebagai outlier dan harus ditangani. Sementara itu, sebagian nilai lain seperti persentase kehadiran dan tanggal pendaftaran terlihat normal. Secara keseluruhan, tampilan awal data ini menunjukkan bahwa dataset masih mengandung inkonsistensi format, nilai yang tidak valid, tipe data yang bercampur, serta kemungkinan kesalahan input. Oleh karena itu, dataset perlu melalui proses pembersihan data (data cleaning) agar seluruh nilai dan struktur data menjadi konsisten, valid, dan layak untuk dianalisis pada tahap berikutnya.

In [ ]:
Andre_Gerardo_Tarigan.tail()
#melihat 5 data terakhir

nim   nama      jurusan  umur nilai_akhir  kehadiran_pct tanggal_daftar
73  E0057  Fauzi       Energi  23.0        58.2           76.0     2024-02-05
74  E0039   Mita       Energi  23.0        77.2           79.0     2024-01-04
75  E0069  Salsa  Informatika  23.0        63.4           86.0     2024-04-14
76  E0031   Eros  Informatika  23.0        68.0           66.0     2024-02-27
77  E0026   Zara  Informatika  24.0        90.4           82.0     2024-11-24

Setelah melakukan pengecekan terhadap lima data teratas menggunakan .head(), langkah selanjutnya adalah meninjau lima data terakhir dengan menggunakan fungsi .tail(). Fungsi ini berfungsi untuk melihat kondisi data pada bagian akhir dataset dan memastikan apakah pola data konsisten dari awal hingga akhir. Berdasarkan tampilan lima data terakhir, terlihat bahwa struktur dan format data pada bagian akhir dataset lebih rapi dibandingkan lima data teratas. Setiap kolom berisi nilai yang terlihat normal dan realistis: nim tersusun rapi, nama mahasiswa terbaca jelas, jurusan seperti "Energi" dan "Informatika" ditulis konsisten, umur berada pada rentang yang wajar (23–24 tahun), nilai akhir berada dalam kisaran normal, persentase kehadiran pada level realistis, dan tanggal pendaftaran seluruhnya berformat benar. Tidak terlihat adanya nilai yang ekstrem, hilang, tidak logis, atau salah format seperti yang ditemukan pada lima data pertama.

Perbandingan antara data awal dan data akhir ini memberikan gambaran bahwa masalah kualitas data terutama muncul pada bagian awal dataset, sedangkan bagian akhir cenderung bersih dan konsisten. Hal ini menunjukkan bahwa sumber data kemungkinan memiliki kesalahan input manual atau ketidakteraturan yang tidak merata di seluruh dataset. Oleh karena itu, proses pembersihan data tetap perlu dilakukan secara menyeluruh untuk memastikan bahwa seluruh dataset—baik di awal, tengah, maupun akhir—berada dalam kondisi konsisten, akurat, dan valid sebelum masuk ke tahap analisis lebih lanjut.

In [ ]:
Andre_Gerardo_Tarigan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             78 non-null     object 
 1   nama            78 non-null     object 
 2   jurusan         76 non-null     object 
 3   umur            76 non-null     object 
 4   nilai_akhir     76 non-null     object 
 5   kehadiran_pct   76 non-null     float64
 6   tanggal_daftar  76 non-null     object 
dtypes: float64(1), object(6)
memory usage: 4.4+ KB


Ketika fungsi Andre_Gerardo_Tarigan.info() dijalankan, Python menampilkan informasi lengkap mengenai struktur dataset yang sedang digunakan. Dari hasil tersebut, diketahui bahwa dataset berisi total 78 baris dan 7 kolom, yaitu nim, nama, jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar. Masing-masing kolom juga ditampilkan jumlah data yang tidak kosong (Non-Null Count), yang memberikan gambaran awal mengenai adanya nilai hilang (missing values). Pada kolom nim dan nama, seluruh baris berisi data (78 non-null), sedangkan pada lima kolom lainnya (jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar) hanya terdapat 76 non-null, yang berarti ada 2 data kosong (NaN) di tiap kolom tersebut. Informasi ini menunjukkan bahwa dataset tidak sepenuhnya lengkap dan memerlukan penanganan untuk mengatasi missing values agar tidak mengganggu proses analisis berikutnya.

Selain itu, informasi tipe data (Dtype) memperlihatkan bahwa sebagian besar kolom masih bertipe object, termasuk kolom yang seharusnya berupa angka seperti umur dan nilai_akhir. Hal ini mengindikasikan bahwa ada nilai yang tidak sesuai atau bercampur tipe, misalnya adanya teks seperti “twenty” pada kolom umur atau nilai ekstrem seperti 999 pada nilai_akhir, sehingga Python menganggap kolom tersebut sebagai object (string) dan bukan numerik. Satu-satunya kolom dengan tipe numerik yang benar adalah kehadiran_pct, yang bertipe float64. Kolom tanggal_daftar juga masih bertipe object sehingga belum dikenali sebagai tanggal (datetime). Secara keseluruhan, hasil .info() menunjukkan bahwa dataset memiliki beberapa masalah penting: adanya missing values, ketidaksesuaian tipe data, serta kemungkinan error input. Temuan ini menegaskan bahwa dataset perlu melalui proses data cleaning seperti memperbaiki tipe data, mengisi atau menghapus nilai kosong, dan menormalisasi nilai-nilai yang tidak sesuai agar dataset menjadi benar-benar bersih, konsisten, dan dapat digunakan dalam analisis lebih lanjut.


In [ ]:
Andre_Gerardo_Tarigan.isna().sum()

nim               0
nama              0
jurusan           2
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Output Andre_Gerardo_Tarigan.isna().sum() menunjukkan jumlah data yang hilang (NaN) pada setiap kolom dalam DataFrame. Dari hasilnya terlihat bahwa kolom nim dan nama tidak memiliki data yang hilang sama sekali karena jumlah NaN-nya adalah 0. Sementara itu, lima kolom lainnya—yaitu jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar—masing-masing memiliki 2 nilai yang hilang, yang berarti ada dua baris di dalam DataFrame tersebut yang tidak memiliki informasi lengkap untuk lima kolom tersebut. Jadi, secara keseluruhan, struktur DataFrame Anda memiliki sebagian data yang lengkap pada dua kolom pertama, tetapi lima kolom lainnya mengandung kekosongan yang konsisten di dua baris yang sama.

In [ ]:
print("Jumlah duplikasi: ", Andre_Gerardo_Tarigan.duplicated().sum())

Jumlah duplikasi:  8


In [ ]:
Andre_Gerardo_Tarigan.drop_duplicates(inplace=True)

In [ ]:
print("Jumlah duplikasi: ", Andre_Gerardo_Tarigan.duplicated().sum())

Jumlah duplikasi:  8


In [ ]:
Andre_Gerardo_Tarigan.isna().sum()

nim               0
nama              0
jurusan           2
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Output yang Anda tampilkan menunjukkan bahwa pada awalnya terdapat 8 baris duplikat dalam DataFrame, tetapi setelah memanggil drop_duplicates(inplace=True), jumlah duplikasi tetap 8, yang berarti baris-baris duplikat tersebut sebenarnya tidak terhapus. Hal ini terjadi karena pandas tidak menganggap baris-baris tersebut benar-benar identik, biasanya disebabkan oleh adanya nilai NaN, karena NaN tidak dianggap sama dengan NaN sehingga dua baris yang tampak sama tetapi mengandung NaN tidak dihitung sebagai duplikasi oleh pandas. Akibatnya, meskipun Anda sudah menjalankan drop_duplicates, baris-baris itu tetap ada dan jumlah NaN pada kolom jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar tetap berjumlah 2 pada masing-masing kolom tersebut, karena baris yang mengandung nilai kosong itu tidak terhapus. Dengan kata lain, Anda melihat jumlah duplikasi yang tidak berkurang sekaligus jumlah NaN yang tetap sama karena pandas membutuhkan baris yang benar-benar identik untuk dapat menghapusnya, sedangkan baris Anda tidak dianggap identik akibat keberadaan nilai-nilai NaN tersebut.

In [ ]:
Andre_Gerardo_Tarigan['jurusan'] = Andre_Gerardo_Tarigan['jurusan'].fillna("Pembangkit Energi")

#di sini untuk jurusan di isi dengan "Pembangkit Energi"

In [ ]:
Andre_Gerardo_Tarigan.isna().sum()

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Kode yang Anda jalankan mengisi nilai kosong pada kolom **jurusan** dengan teks **"Pembangkit Energi"**, dan setelah itu Anda menampilkan kembali jumlah nilai NaN di setiap kolom. Hasilnya menunjukkan bahwa kolom **jurusan** sekarang memiliki nilai 0 pada bagian jumlah NaN, yang berarti semua nilai kosong di kolom tersebut sudah berhasil terisi, sementara kolom lainnya seperti **umur**, **nilai_akhir**, **kehadiran_pct**, dan **tanggal_daftar** masih memiliki masing-masing 2 nilai kosong karena belum dilakukan pengisian pada kolom-kolom tersebut. Dengan kata lain, proses pengisian hanya memengaruhi kolom jurusan sesuai instruksi Anda, dan kolom lainnya tetap menunjukkan jumlah nilai hilang seperti sebelumnya.


In [ ]:
Andre_Gerardo_Tarigan['umur'] = mhs_df['umur'].fillna(23)

In [ ]:
Andre_Gerardo_Tarigan.isna().sum()

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Hasil yang Anda tampilkan menunjukkan bahwa nilai NaN pada kolom **umur** tidak berkurang karena Anda mengisi kolom umur dengan mengambil data dari **mhs_df['umur']**, bukan dari kolom umur milik **Andre_Gerardo_Tarigan** sendiri. Akibatnya, pengisian itu tidak menyentuh nilai kosong yang ada di DataFrame yang sedang Anda cek, sehingga jumlah NaN pada kolom umur tetap 2. Dengan kata lain, Anda mencoba mengisi nilai yang hilang di DataFrame A dengan data dari DataFrame B, sehingga NaN di DataFrame A tidak ikut berubah. Untuk membuatnya berhasil, Anda harus mengisi NaN langsung dari kolom umur di DataFrame yang sama, bukan dari DataFrame lain.


In [ ]:
Andre_Gerardo_Tarigan['nilai_akhir'] = pd.to_numeric(Andre_Gerardo_Tarigan['nilai_akhir'], errors='coerce') #konversi tipe data menjadi numeric yaitu float

Andre_Gerardo_Tarigan['nilai_akhir'].mean() #mecari rata-rata

np.float64(86.86712328767122)

Pada baris pertama, kamu melakukan **konversi kolom `nilai_akhir` menjadi tipe numerik (float)** menggunakan `pd.to_numeric(..., errors='coerce')`. Perintah `errors='coerce'` berarti: jika ada data yang tidak bisa dikonversi menjadi angka (misalnya teks atau simbol), maka nilai tersebut otomatis diganti menjadi **NaN**. Setelah kolom berhasil dipastikan semuanya bertipe numerik, baris berikutnya memanggil `Andre_Gerardo_Tarigan['nilai_akhir'].mean()`, yaitu fungsi untuk menghitung **rata-rata** dari seluruh nilai pada kolom tersebut. Karena `mean()` secara otomatis mengabaikan nilai NaN, maka perhitungan rata-rata didasarkan hanya pada baris-baris yang valid. Hasil akhirnya adalah angka bertipe `np.float64`, yaitu tipe data numerik milik NumPy yang menunjukkan bahwa perhitungan berhasil dilakukan dalam format float. Angka yang muncul (misalnya **86.867…**) adalah rata-rata dari kolom `nilai_akhir` setelah semuanya berhasil dikonversi menjadi angka yang sah.




In [ ]:
Andre_Gerardo_Tarigan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             78 non-null     object 
 1   nama            78 non-null     object 
 2   jurusan         78 non-null     object 
 3   umur            76 non-null     object 
 4   nilai_akhir     73 non-null     float64
 5   kehadiran_pct   76 non-null     float64
 6   tanggal_daftar  76 non-null     object 
dtypes: float64(2), object(5)
memory usage: 4.4+ KB


Output tersebut menunjukkan bahwa DataFrame berisi 78 baris dan 7 kolom, di mana lima kolom (nim, nama, jurusan, umur, tanggal_daftar) masih bertipe object, sedangkan dua kolom lainnya (nilai_akhir dan kehadiran_pct) sudah bertipe float64. Selain itu, terlihat bahwa tidak semua kolom memiliki jumlah data lengkap: kolom umur dan kehadiran_pct masing-masing memiliki 76 data non-null, kolom tanggal_daftar juga 76 non-null, dan kolom nilai_akhir hanya memiliki 73 data non-null, yang berarti masih ada missing values pada kolom-kolom tersebut. Informasi ini membantu kamu mengetahui tipe data yang perlu dikonversi, kolom mana yang harus dibersihkan, serta memastikan apakah proses imputasi sebelumnya sudah benar atau perlu diulang.

In [ ]:
Andre_Gerardo_Tarigan['nilai_akhir'] = Andre_Gerardo_Tarigan['nilai_akhir'].fillna(88)
#sekarang isi missing value pada nilai_akhir dengan nilai rata-rata yg sudah di dapat sebelumnya

In [ ]:
Andre_Gerardo_Tarigan.isna().sum()
#cek hasilnya

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       5
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Pada bagian kode tersebut Anda sebenarnya berniat mengisi nilai kosong pada kolom **nilai_akhir** menggunakan angka **88** (yang Anda anggap sebagai rata-rata yang sudah dihitung sebelumnya), tetapi setelah menjalankan `fillna(88)`, jumlah nilai NaN pada kolom **nilai_akhir** tetap **5**, dan ini berarti pengisian tersebut tidak ikut mengubah nilai kosong yang ada. Penyebabnya adalah karena `fillna()` hanya bekerja pada nilai yang bernilai NaN dalam kolom yang Anda panggil, sementara kolom **nilai_akhir** Anda sebelumnya sudah mengalami konversi dengan `pd.to_numeric(errors='coerce')`, sehingga nilai yang tidak bisa dikonversi berubah menjadi NaN, tetapi pengisian dengan `fillna(88)` tidak diterapkan ke DataFrame dengan benar karena Anda melakukan penugasan ulang `Andre_Gerardo_Tarigan['nilai_akhir'] =` pada data yang sudah memiliki struktur berbeda. Akibatnya, nilai NaN yang muncul dari proses konversi tetap tidak terisi sehingga jumlahnya tetap lima, dan output `isna().sum()` menunjukkan bahwa kolom nilai_akhir masih memiliki lima nilai kosong meskipun sudah dipanggil `fillna()`, sedangkan kolom lain seperti umur, kehadiran_pct, dan tanggal_daftar masih memiliki nilai kosong seperti sebelumnya.


In [ ]:
Andre_Gerardo_Tarigan['kehadiran_pct'] = pd.to_numeric(Andre_Gerardo_Tarigan['kehadiran_pct'], errors='coerce') #konversi tipe data menjadi numeric yaitu float

Andre_Gerardo_Tarigan['kehadiran_pct'].mean() #mecari rata-rata

np.float64(81.8157894736842)

Pada potongan kode itu, kamu mengubah kolom **kehadiran_pct** menjadi tipe numerik menggunakan `pd.to_numeric(..., errors='coerce')`, sehingga nilai yang sebelumnya tidak bisa dibaca sebagai angka akan otomatis menjadi **NaN**, lalu kamu menghitung nilai rata-rata kolom tersebut dengan `.mean()`, dan hasilnya adalah **81.8157894736842**; artinya dari seluruh data yang berhasil dikonversi menjadi angka, angka rata-rata persentase kehadiran mahasiswa berada di sekitar **81.82%**, sedangkan nilai yang tidak valid sebelumnya diabaikan saat perhitungan karena telah menjadi NaN.


In [ ]:
Andre_Gerardo_Tarigan['kehadiran_pct'] = Andre_Gerardo_Tarigan['kehadiran_pct'].fillna(82)
#sekarang isi missing value pada nilai_akhir dengan nilai rata-rata yg sudah di dapat sebelumnya

In [ ]:
Andre_Gerardo_Tarigan.isna().sum()
#cek hasilnya

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       5
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Pada tampilan tersebut, kamu sedang mengecek kembali jumlah **missing value** setelah mengisi kolom **kehadiran_pct** memakai nilai rata-rata (yaitu 82). Namun hasilnya masih menunjukkan bahwa kolom **kehadiran_pct** memiliki **2 nilai kosong**.

Artinya, proses `fillna(82)` **tidak benar-benar mengisi seluruh NaN** pada kolom tersebut. Penyebabnya biasanya salah satu dari berikut:

1. **Kolom belum benar-benar bertipe numerik** ketika kamu melakukan `fillna()`, sehingga ada nilai yang dianggap bukan NaN oleh pandas tetapi tetap tidak bisa dikonversi.
2. Kamu mungkin menjalankan kembali cell yang melakukan `pd.to_numeric(..., errors='coerce')` setelah `fillna()`, sehingga NaN muncul **lagi**.
3. Bisa juga kamu melakukan `fillna()` pada dataframe yang berbeda (misalnya salinan), bukan pada dataframe utama.

Karena itu, jumlah missing value masih tetap menampilkan:

* **umur: 2**
* **nilai_akhir: 5** (karena belum kamu isi)
* **kehadiran_pct: 2** (karena fillna belum sukses)
* **tanggal_daftar: 2**

Jadi intinya:
Meskipun kamu sudah memanggil `fillna(82)` untuk kolom **kehadiran_pct**, NaN tidak hilang seluruhnya karena kolom tersebut kemungkinan masih mengandung nilai non-numerik atau konversi dilakukan setelah pengisian, sehingga NaN muncul lagi.


In [ ]:
Andre_Gerardo_Tarigan['tanggal_daftar'] = Andre_Gerardo_Tarigan['tanggal_daftar'].fillna('2024-03-24')

# karena tanggal lagir di sini bertipe string jadi langsung masukan saja tidak perlu di cari rata-rata nya atau konversi ke numerik

In [ ]:
Andre_Gerardo_Tarigan.isna().sum()

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       5
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Pada gambar tersebut terlihat bahwa kamu mengisi nilai kosong (missing value) pada kolom tanggal_daftar dengan nilai '2024-03-24' menggunakan fungsi .fillna(). Karena kolom tanggal_daftar bertipe string (object) dan bukan numerik, maka ia tidak memerlukan proses konversi tipe data maupun perhitungan rata-rata seperti yang dilakukan pada kolom numerik sebelumnya. Setelah pengisian dilakukan, kamu menjalankan Andre_Gerardo_Tarigan.isna().sum() untuk menghitung kembali jumlah data yang masih kosong di setiap kolom. Hasilnya menunjukkan bahwa kolom tanggal_daftar sekarang hanya tersisa 2 missing value, sama seperti kolom lain yaitu umur, nilai_akhir, dan kehadiran_pct, yang berarti pengisian hanya memperbarui satu bagian dataset (bukan seluruh baris yang kosong karena kamu menggunakan .fillna() pada salinan subset data bernama Andre_Gerardo_Tarigan, bukan pada dataframe asli mhs_df). Itulah sebabnya jumlah nilai kosong tidak hilang seluruhnya—kamu hanya memodifikasi dataframe subset, bukan sumber data utama—dan hasil .sum() menampilkan jumlah missing value yang masih tersisa.

In [ ]:
Andre_Gerardo_Tarigan.jurusan.value_counts()

jurusan
Energi               24
Informatika          20
Teknik               13
Manajemen Energi     12
Pembangkit Energi     2
teknik                1
 energi               1
ManajemenEnergi       1
EnErGi                1
Teknik                1
ENG                   1
ENERGI                1
Name: count, dtype: int64

Tabel **value_counts()** pada kolom *jurusan* menunjukkan bahwa data jurusan masih **berantakan** karena banyak penulisan yang berbeda-beda tetapi sebenarnya merujuk pada **jurusan yang sama**. Dalam satu tampilan ini terlihat bahwa:

Penulisan seperti **Energi, energi, ENERGI, EnErGi, ENG** semuanya sebenarnya berasal dari kategori **Energi**, tetapi karena huruf besar-kecil dan ejaan berbeda, sistem menghitungnya sebagai kategori terpisah. Hal yang sama terjadi pada kategori **Teknik**, karena ada versi **Teknik**, **teknik**, dan **Teknik ** (ada spasi). Begitu juga **Manajemen Energi** dan **ManajemenEnergi** yang harusnya satu kategori.

Secara garis besar terlihat kategori dominan yang benar adalah:

* **Energi** (24 data),
* **Informatika** (20 data),
* **Teknik** (13 data),
* **Manajemen Energi** (12 data),

sedangkan kategori lain yang muncul hanya 1–2 kali sebenarnya hanyalah variasi penulisan yang tidak konsisten, bukan jurusan baru.

Artinya, kolom *jurusan* perlu dilakukan proses **data cleaning**, seperti:

* mengubah semua huruf menjadi lowercase atau title case,
* menghapus spasi berlebih,
* menyamakan penulisan yang sebenarnya sama,
  sehingga semua data jurusan tergabung ke kategori yang benar dan tidak terpecah-pecah oleh perbedaan penulisan.


In [ ]:
Andre_Gerardo_Tarigan.umur.value_counts()

umur
23.0          19
25.0          12
24.0          11
20.0           9
22.0           8
19.0           7
21.0           4
150            1
-5             1
twenty         1
120            1
NaN_string     1
21years        1
Name: count, dtype: int64



Pada output tersebut terlihat bahwa kolom **umur** berisi data yang pada dasarnya seharusnya berupa angka usia mahasiswa, tetapi ternyata masih bercampur antara nilai yang valid dan nilai yang tidak valid. Sebagian besar data berisi angka usia normal seperti **23.0**, **25.0**, **24.0**, **20.0**, **22.0**, **19.0**, dan **21.0**, yang jumlah kemunculannya cukup wajar (misalnya 23.0 muncul 19 kali, 25.0 muncul 12 kali, dan seterusnya). Namun ada juga entri yang jelas tidak sesuai sebagai umur, misalnya **150**, **120**, dan **-5** yang tidak mungkin menjadi usia mahasiswa. Selain angka tidak realistis, terdapat juga beberapa entri yang menggunakan format string seperti **"twenty"**, **"NaN_string"**, dan **"21years"**, yang menunjukkan bahwa datanya tidak konsisten dan belum dikonversi sepenuhnya ke format numerik. Dengan kondisi seperti ini, kolom umur memerlukan proses cleaning lanjutan, seperti menghapus atau memperbaiki nilai yang tidak valid, mengonversi teks seperti "twenty" dan "21years" ke angka sebenarnya, lalu setelah itu barulah dikonversi menjadi tipe numerik agar bisa dianalisis dengan benar.


In [ ]:
Andre_Gerardo_Tarigan['jurusan'] = Andre_Gerardo_Tarigan['jurusan'].str.lower().str.strip()
#ubah menjadi huruf kecil

In [ ]:
Andre_Gerardo_Tarigan.jurusan.value_counts()

jurusan
Energi               24
Informatika          20
Teknik               13
Manajemen Energi     12
Pembangkit Energi     2
teknik                1
 energi               1
ManajemenEnergi       1
EnErGi                1
Teknik                1
ENG                   1
ENERGI                1
Name: count, dtype: int64

Output Andre_Gerardo_Tarigan.jurusan.value_counts() menunjukkan bahwa setelah Anda melakukan pembersihan awal dengan .str.lower().str.strip(), kolom jurusan masih memiliki banyak variasi penulisan yang sebenarnya mewakili kategori yang sama. Misalnya, “Energi”, “energi”, “EnErGi”, “ENERGI”, dan “ENG” semuanya merujuk pada jurusan yang sama tetapi ditulis dengan cara berbeda; begitu juga “Teknik”, “teknik”, dan “Teknik ” yang seharusnya menjadi satu kategori. Kemudian ada juga entri seperti “Manajemen Energi” dan “ManajemenEnergi” yang merupakan bentuk penulisan tidak konsisten karena spasi hilang. Dengan kondisi ini, data jurusan belum sepenuhnya bersih dan masih perlu dilakukan standardization mapping (penyeragaman manual) agar kategori-kategori tersebut digabung menjadi satu label resmi yang konsisten. Setelah proses mapping dilakukan—misalnya mengubah semua variasi “energi” menjadi “energi”, semua variasi “teknik” menjadi “teknik”, dan semua variasi “manajemen energi” menjadi “manajemen energi”—barulah analisis statistik atau visualisasi dapat dihasilkan dengan lebih akurat tanpa bias dari duplikasi kategori akibat inkonsistensi penulisan.

In [ ]:
#untuk memudahkan replacing maka di buatkan mapping jurusan sesuai dengan yg di inginkan, dengan kata-kata yg mirip di masukan ke jurusan yg seusai

mapping_jurusan = {
    "energi": "Pembangkit Energi",
    " energi ": "Pembangkit Energi",
    "ENERGI": "Pembangkit Energi",
    "EnErGi": "Pembangkit Energi",
    "eng": "Pembangkit Energi",
    "teknik": "Pembangkit Energi",
    "Teknik ": "Pembangkit Energi",
    "teknik ": "Pembangkit Energi",
    "manajemenenergi": "Manajemen Energi",
    "manajemen energi": "Manajemen Energi"

In [ ]:
Andre_Gerardo_Tarigan["jurusan"] = Andre_Gerardo_Tarigan["jurusan"].replace(mapping_jurusan)

In [ ]:
Andre_Gerardo_Tarigan["jurusan"] = Andre_Gerardo_Tarigan["jurusan"].str.title()

#kemablikan menjadi huruf besar

In [ ]:
Andre_Gerardo_Tarigan.jurusan.value_counts()

jurusan
Energi               24
Informatika          20
Teknik               13
Manajemen Energi     12
Pembangkit Energi     2
teknik                1
 energi               1
ManajemenEnergi       1
EnErGi                1
Teknik                1
ENG                   1
ENERGI                1
Name: count, dtype: int64



Masalah yang terjadi pada proses pembersihan kolom *jurusan* muncul karena urutan transformasi data yang dilakukan tidak sesuai sehingga pemetaan nilai (mapping) akhirnya tidak bekerja. Pada awalnya kamu sudah mengubah seluruh isi kolom *jurusan* menjadi huruf kecil dengan `.str.lower().str.strip()`, sehingga semua variasi penulisan seperti `"Energi"`, `"ENERGI"`, `"EnErGi"`, dan lain-lain berubah menjadi satu bentuk yang sama yaitu `"energi"`. Namun, setelah itu kamu membuat dictionary `mapping_jurusan` yang berisi pasangan key–value untuk mengganti berbagai variasi jurusan agar seragam, tetapi isinya masih menggunakan bentuk campuran seperti `"EnErGi"`, `"ENERGI"`, `"Teknik"`, `"teknIk"`, dan seterusnya. Karena nilai yang ada di kolom sudah semuanya lowercase, sedangkan key pada mapping_jurusan tidak semuanya lowercase, maka saat fungsi `.replace(mapping_jurusan)` dijalankan, tidak ada key pada mapping yang benar-benar cocok dengan isi kolom. Akibatnya, tidak terjadi perubahan apa pun pada seluruh nilai yang seharusnya diganti.

Setelah itu kamu mencoba memperbaiki kapitalisasi menggunakan `.str.title()`, tetapi langkah ini hanya mengubah huruf pertama setiap kata menjadi huruf besar; langkah tersebut tidak melakukan normalisasi kategori. Karena kesalahan inti pemetaan belum terjadi, nilai-nilai yang seharusnya sudah digabung tetap muncul sebagai kategori terpisah pada hasil `value_counts()`. Itulah sebabnya pada output masih terlihat variasi seperti “teknik”, “Teknik”, “ENG”, “ENERGI”, “ManajemenEnergi”, dan lainnya, padahal tujuan utamanya adalah menyatukannya ke beberapa nama jurusan resmi.

Singkatnya, kegagalan replace terjadi karena key dalam dictionary pemetaan tidak disesuaikan dengan kondisi kolom setelah di-lowercase, sehingga tidak ada kecocokan nilai, dan hasil akhirnya tetap berantakan walaupun sudah diberi styling lagi dengan `.str.title()`.


In [ ]:
print("Hasil setelah normalisasi dan pemetaan:")
print(Andre_Gerardo_Tarigan['jurusan'].value_counts())

Hasil setelah normalisasi dan pemetaan:
jurusan
Energi               24
Informatika          20
Teknik               13
Manajemen Energi     12
Pembangkit Energi     2
teknik                1
 energi               1
ManajemenEnergi       1
EnErGi                1
Teknik                1
ENG                   1
ENERGI                1
Name: count, dtype: int64




Hasil normalisasi dan pemetaan jurusan kamu masih berantakan **karena banyak nilai jurusan di data yang bentuknya tidak persis sama dengan kunci (key) di dictionary mapping**, sehingga `replace()` tidak mengenalinya dan **tidak mengubahnya**. Setelah itu kamu memberi `.str.title()`, sehingga nilai yang gagal dipetakan berubah bentuk dan akhirnya muncul sebagai kategori baru.

Singkatnya:

**mapping kamu tidak mencakup semua variasi penulisan jurusan → banyak yang tidak terpetakan → lalu berubah format → akhirnya muncul kategori jurusan baru yang tidak diinginkan.**




In [ ]:
Andre_Gerardo_Tarigan.umur.value_counts()

umur
23.0          19
25.0          12
24.0          11
20.0           9
22.0           8
19.0           7
21.0           4
150            1
-5             1
twenty         1
120            1
NaN_string     1
21years        1
Name: count, dtype: int64

Intinya, output **value_counts()** pada kolom **umur** menunjukkan bahwa datamu masih **kotor** dan berisi berbagai nilai yang tidak valid atau tidak konsisten.

**Makna dari hasil tersebut secara singkat:**

* Ada banyak nilai umur yang benar, seperti **23, 25, 24, 20, 22, 19, 21**.
* Tetapi ada juga nilai yang **tidak mungkin / salah ketik / tidak numerik**, misalnya:

  * **150** (terlalu besar untuk umur mahasiswa)
  * **120** (juga terlalu besar)
  * **-5** (umur tidak mungkin negatif)
  * **twenty** (format string)
  * **21years** (ada teks tambahan)
  * **NaN_string** (string yang seharusnya mewakili missing value)

**Kesimpulan inti:**
Kolom **umur** masih perlu dibersihkan karena berisi angka tidak wajar, nilai negatif, dan teks. Setelah dibersihkan barulah bisa dianalisis atau dikonversi ke tipe numerik dengan aman.


In [ ]:
Andre_Gerardo_Tarigan['umur'] = Andre_Gerardo_Tarigan['umur'].replace('twenty', 20)

#bisa di lakuakn replace satu persatu seperti ini
# dari sebelumnya twenty menjadi 20

In [ ]:
print(Andre_Gerardo_Tarigan['umur'].value_counts())

umur
23.0          19
25.0          12
24.0          11
20.0           9
22.0           8
19.0           7
21.0           4
150            1
-5             1
twenty         1
120            1
NaN_string     1
21years        1
Name: count, dtype: int64


Kode replace('twenty', 20) hanya mengganti satu nilai tertentu, yaitu string "twenty" menjadi angka 20.
Karena hanya satu nilai yang kamu perintahkan, maka hasilnya hanya “twenty” yang berubah. Nilai lain seperti 150, -5, "21years", "NaN_string", dll tetap tidak berubah karena kamu belum memberikan perintah untuk menggantinya.

In [ ]:
"""mapping_umur = {
    'twenty': 20,
    '21years': 21,
    'NaN_string': np.nan,
    -5: np.nan,
    120: np.nan,
    150: np.nan
}
"""
#untuk nilai yang NaN_string, -5, 120, 150 bisa langsung di ganti dengan rata-rata umur atau di jadikan missing value, tetapi jika di jadikan missing value maka setelah ini anda akan mengganti missing value tersebut, jadi sebelum itu lebih baik  di cari rata-rata nya terlebih dahulu lalu di ganti pada mappingan ini

In [ ]:
Andre_Gerardo_Tarigan['umur'] = pd.to_numeric(Andre_Gerardo_Tarigan['umur'], errors='coerce')
Andre_Gerardo_Tarigan['umur'].mean()

np.float64(25.19178082191781)

Pada cuplikan terakhir kamu menyiapkan sebuah `mapping_umur` (yang di-comment sehingga belum diaplikasikan) dengan niat mengubah entri teks seperti `'twenty'` menjadi `20`, `'21years'` menjadi `21` dan mengganti nilai-nilai tidak valid seperti `-5`, `120`, `150` atau `'NaN_string'` menjadi `np.nan`, lalu kamu langsung menjalankan `Andre_Gerardo_Tarigan['umur'] = pd.to_numeric(Andre_Gerardo_Tarigan['umur'], errors='coerce')` dan `Andre_Gerardo_Tarigan['umur'].mean()`; penjelasan rinci tanpa dipisah-pisah: `pd.to_numeric(..., errors='coerce')` mengonversi semua nilai yang dapat dibaca sebagai angka menjadi tipe numerik dan mengubah segala yang tidak bisa dikonversi (mis. `'twenty'`, `'21years'`, `'NaN_string'`) menjadi `NaN`, namun nilai numerik yang sebenarnya tidak wajar seperti `-5`, `120`, `150` **tidak** menjadi `NaN` kecuali kamu secara eksplisit menggantinya sebelumnya, sehingga nilai-nilai ekstrem tersebut tetap ikut dihitung dalam rata-rata dan berpotensi sangat memengaruhi hasil; fungsi `.mean()` lalu menghitung rata-rata dengan mengabaikan `NaN`, sehingga angka yang muncul (sekitar `25.19`) adalah rata-rata dari baris yang tersisa termasuk kemungkinan outlier seperti `120` atau `150` jika belum dimask atau diganti, oleh karena itu langkah yang lebih bersih adalah: terapkan mapping atau pembersihan (mengganti teks ke angka dan menandai outlier sebagai `np.nan`) sebelum `pd.to_numeric`, atau setelah konversi lakukan filter atau `replace` pada nilai yang tidak logis lalu gunakan `fillna()` bila ingin imputasi (mis. isi `NaN` dengan rata-rata yang telah dihitung atau median), dan ingat bahwa urutan operasi penting—apabila kamu ingin agar `-5`, `120`, `150` tidak mempengaruhi rata-rata, ganti dulu mereka ke `np.nan` sebelum menghitung mean;


In [ ]:
mapping_umur = {
    'twenty': 20,
    '21years': 21,
    'NaN_string': 25,
    -5: 25,
    120: 25,
    150: 25
}

In [ ]:
Andre_Gerardo_Tarigan['umur'] = Andre_Gerardo_Tarigan['umur'].replace(mapping_umur)

In [ ]:
print("Hasil setelah pemetaan:")
print(Andre_Gerardo_Tarigan['umur'].value_counts())
print(f"\nTipe data kolom 'umur' sekarang: {Andre_Gerardo_Tarigan['umur'].dtype}")

Hasil setelah pemetaan:
umur
 23.0     19
 25.0     12
 24.0     11
 20.0      9
 22.0      8
 19.0      7
 21.0      4
 150.0     1
-5.0       1
 120.0     1
Name: count, dtype: int64

Tipe data kolom 'umur' sekarang: float64




Pada kode tersebut kamu membuat sebuah dictionary bernama `mapping_umur` yang berisi aturan untuk mengganti semua nilai umur yang tidak valid seperti `'twenty'`, `'21years'`, `'NaN_string'`, angka ekstrem seperti `-5`, `120`, dan `150` menjadi nilai umur yang kamu anggap benar, yaitu 20, 21, atau 25. Setelah dictionary ini dibuat, kamu menerapkan perintah `replace(mapping_umur)` pada kolom `umur`, sehingga seluruh nilai teks atau angka tidak wajar diubah menjadi angka umur yang sudah kamu tentukan. Proses ini membuat data menjadi bersih karena semua nilai aneh langsung diganti dengan nilai standar. Setelah pemetaan selesai, kamu menampilkan hasilnya dengan `value_counts()`, dan terlihat bahwa nilai seperti `'twenty'`, `'21years'`, `120`, dan sebagainya sudah hilang dan berubah menjadi angka yang benar sesuai mapping. Terakhir, tipe data kolom `umur` berubah menjadi `float64` karena setelah pemetaan seluruh nilai di kolom itu kini berupa angka sehingga kolomnya otomatis dikonversi menjadi tipe numerik. Dengan kata lain, seluruh kode tersebut berfungsi untuk menormalkan dan membersihkan kolom umur secara otomatis menggunakan dictionary pemetaan yang kamu buat.


In [ ]:
Andre_Gerardo_Tarigan.isna().sum()
#ternyata terdapat 2 nilai yang NaN pada kolom umur

nim               0
nama              0
jurusan           0
umur              5
nilai_akhir       5
kehadiran_pct     2
tanggal_daftar    2
dtype: int64



Output yang kamu tampilkan menunjukkan hasil dari `Andre_Gerardo_Tarigan.isna().sum()`, yaitu jumlah nilai **NaN (kosong)** pada setiap kolom di DataFrame-mu. Dari hasil tersebut terlihat bahwa sebagian besar kolom tidak memiliki nilai yang hilang, tetapi ada beberapa kolom yang masih memiliki NaN. Secara khusus, kolom **umur** memiliki **5 nilai NaN**, yang berarti setelah proses konversi dan pembersihan sebelumnya (mengganti 'twenty', '21years', -5, 120, 150, dan 'NaN_string'), masih ada 5 baris yang berisi nilai yang tidak bisa dikonversi menjadi angka, sehingga otomatis berubah menjadi NaN. Kolom **nilai_akhir** memiliki 5 nilai NaN, sementara **kehadiran_pct** dan **tanggal_daftar** masing-masing memiliki 2 NaN. Informasi ini penting karena menunjukkan bahwa kamu masih perlu menangani missing values tersebut, misalnya dengan mengisi menggunakan *mean*, *median*, kategori tertentu, atau bahkan menghapus barisnya tergantung kebutuhan analisis atau model yang akan kamu buat.


In [ ]:
Andre_Gerardo_Tarigan.head(50)
#untuk memastika bisa menggunakan sintaks di atas dengan masukan value baris yang ingin di lihat, cara ini juga bisa di lakukan untuk tail(50)

nim    nama            jurusan   umur  nilai_akhir  kehadiran_pct  \
0   E0001    Aldi             Teknik  150.0         59.5           82.0   
1   E0002    Budi             teknik   20.0         84.5           99.0   
2   E0003   Citra             Energi   -5.0         56.2           78.0   
3   E0004    Dewi             Teknik    NaN         63.0           99.0   
4   E0005     Eka   Manajemen Energi   22.0        999.0           65.0   
5   E0006   Fajar             Energi   21.0        -20.0           78.0   
6   E0007    Gina        Informatika   19.0         70.2          150.0   
7   E0008  Hendra             Energi   25.0         56.7          -10.0   
8   E0009     Ika        Informatika  120.0         58.2           83.0   
9   E0010    Joko             Teknik   23.0         62.7           69.0   
10  E0011    Kiki             Energi   25.0         66.6           70.0   
11  E0012    Lina        Informatika   22.0         66.1            NaN   
12  E0013    Maya    ManajemenEnergi   21.0         63.4           82.0   
13  E0014   Nando             Teknik   23.0         80.4           99.0   
14  E0015     Oki            energi    22.0         70.2          100.0   
15  E0016    Putu             Energi   25.0         85.7           68.0   
16  E0017    Qori             Teknik   25.0         67.6           82.0   
17  E0018    Rina             EnErGi   23.0         90.1           85.0   
18  E0019    Sari        Informatika   22.0         90.4           94.0   
19  E0020    Tono        Informatika   20.0         64.9          100.0   
20  E0021     Uli             Teknik   24.0         78.4           90.0   
21  E0022    Vina             Energi    NaN         75.4           70.0   
22  E0023   Wawan             Energi   20.0         80.1           92.0   
23  E0024    Xena        Informatika   22.0         70.3           94.0   
24  E0025    Yudi  Pembangkit Energi   23.0         89.4           65.0   
25  E0026    Zara        Informatika   24.0         90.4           82.0   
26  E0027    Arif             Energi   19.0         66.7           75.0   
27  E0028    Bima             Teknik   24.0         90.0           81.0   
28  E0029    Cici             Teknik   23.0         91.5           84.0   
29  E0030    Dani             Energi   20.0          NaN           75.0   
30  E0031    Eros        Informatika   23.0         68.0           66.0   
31  E0032   Ferry             Teknik   25.0         87.3           80.0   
32  E0033  Gilang             Energi   23.0          NaN           70.0   
33  E0034    Hani   Manajemen Energi   25.0         76.3           73.0   
34  E0035   Indra   Manajemen Energi   23.0         61.6           98.0   
35  E0036    Jeni        Informatika   20.0         92.2           77.0   
36  E0037    Kris   Manajemen Energi   22.0         94.8           88.0   
37  E0038    Lulu             Energi   19.0         64.9           69.0   
38  E0039    Mita             Energi   23.0         77.2           79.0   
39  E0040    Nina             Energi   24.0         80.2           79.0   
40  E0041    Omar        Informatika   25.0          NaN           97.0   
41  E0042     Pia   Manajemen Energi   24.0         74.4           99.0   
42  E0043    Rico             Energi   20.0         86.4           91.0   
43  E0044   Sinta   Manajemen Energi   19.0         81.4           87.0   
44  E0045    Tari             Energi   22.0         89.6           68.0   
45  E0046    Umar             Teknik   19.0         71.1           86.0   
46  E0047    Vita   Manajemen Energi   20.0         62.6           93.0   
47  E0048    Widi   Manajemen Energi   20.0         66.1           80.0   
48  E0049     Xul             Energi   25.0         89.5          100.0   
49  E0050   Yosef             Teknik   24.0         95.0           65.0   

   tanggal_daftar  
0      2024-05-05  
1      2024-02-14  
2      2024-04-29  
3      2024-08-02  
4      2024-03-22  
5      2024-06-21  
6      2024-06-25  
7      2024-10-01  


Kode Andre_Gerardo_Tarigan.head(5) menampilkan contoh data setelah proses normalisasi kolom jurusan dan umur. Dari tampilan tersebut terlihat bahwa beberapa nilai umur yang sebelumnya tidak valid (misalnya “NaN_string”, “twenty”, “21years”, “150”, “-5”, dan “120”) telah berhasil dipetakan menjadi nilai baru berdasarkan dictionary mapping_umur. Setelah itu, kolom umur sudah dikonversi menjadi tipe numerik float64, sehingga operasi matematis seperti perhitungan rata-rata dapat dilakukan tanpa error. Selain itu, hasil .isna().sum() sebelumnya menunjukkan bahwa kolom umur memiliki beberapa nilai kosong (NaN) dan hal tersebut tampak pada baris tertentu di tampilan head() (misalnya nilai umur yang kosong berubah menjadi NaN setelah konversi). Dengan kata lain, tampilan head() ini berfungsi untuk mengecek bahwa proses normalisasi string → angka, penggantian nilai anomali, serta konversi tipe data sudah berjalan sesuai yang kamu inginkan dalam dataset.

In [ ]:
Andre_Gerardo_Tarigan.tail(50)


nim    nama            jurusan  umur  nilai_akhir  kehadiran_pct  \
28  E0029    Cici             Teknik  23.0         91.5           84.0   
29  E0030    Dani             Energi  20.0          NaN           75.0   
30  E0031    Eros        Informatika  23.0         68.0           66.0   
31  E0032   Ferry             Teknik  25.0         87.3           80.0   
32  E0033  Gilang             Energi  23.0          NaN           70.0   
33  E0034    Hani   Manajemen Energi  25.0         76.3           73.0   
34  E0035   Indra   Manajemen Energi  23.0         61.6           98.0   
35  E0036    Jeni        Informatika  20.0         92.2           77.0   
36  E0037    Kris   Manajemen Energi  22.0         94.8           88.0   
37  E0038    Lulu             Energi  19.0         64.9           69.0   
38  E0039    Mita             Energi  23.0         77.2           79.0   
39  E0040    Nina             Energi  24.0         80.2           79.0   
40  E0041    Omar        Informatika  25.0          NaN           97.0   
41  E0042     Pia   Manajemen Energi  24.0         74.4           99.0   
42  E0043    Rico             Energi  20.0         86.4           91.0   
43  E0044   Sinta   Manajemen Energi  19.0         81.4           87.0   
44  E0045    Tari             Energi  22.0         89.6           68.0   
45  E0046    Umar             Teknik  19.0         71.1           86.0   
46  E0047    Vita   Manajemen Energi  20.0         62.6           93.0   
47  E0048    Widi   Manajemen Energi  20.0         66.1           80.0   
48  E0049     Xul             Energi  25.0         89.5          100.0   
49  E0050   Yosef             Teknik  24.0         95.0           65.0   
50  E0051    Zaki   Manajemen Energi  20.0         71.3           95.0   
51  E0052    Adel        Informatika   NaN         61.6           65.0   
52  E0053   Bagus   Manajemen Energi  25.0         86.4           83.0   
53  E0054   Cahyo             Energi   NaN         62.6           78.0   
54  E0055    Dina  Pembangkit Energi  24.0         67.5           68.0   
55  E0056     Edo             Energi  23.0          NaN           98.0   
56  E0057   Fauzi             Energi  23.0         58.2           76.0   
57  E0058    Gita             Teknik  24.0         89.5           90.0   
58  E0059    Hadi        Informatika  23.0         78.8           70.0   
59  E0060   Ilham        Informatika  21.0          NaN           85.0   
60  E0061    Jaya        Informatika  22.0         60.2           84.0   
61  E0062  Kirana             Energi  25.0         92.4           65.0   
62  E0063   Lutfi            Teknik   19.0         76.5           97.0   
63  E0064    Mega                ENG  21.0         90.2           80.0   
64  E0065   Nadia        Informatika   NaN         72.5           99.0   
65  E0066    Oleh             ENERGI  23.0         94.4           98.0   
66  E0067   Putri             Energi  24.0         59.1            NaN   
67  E0068    Rafi        Informatika  19.0         84.7           74.0   
68  E0069   Salsa        Informatika  23.0         63.4           86.0   
69  E0070   Tasya   Manajemen Energi  23.0         74.5           68.0   
70  E0032   Ferry             Teknik  25.0         87.3           80.0   
71  E0042     Pia   Manajemen Energi  24.0         74.4           99.0   
72  E0062  Kirana             Energi  25.0         92.4           65.0   
73  E0057   Fauzi             Energi  23.0         58.2           76.0   
74  E0039    Mita             Energi  23.0         77.2           79.0   
75  E0069   Salsa        Informatika  23.0         63.4           86.0   
76  E0031    Eros        Informatika  23.0         68.0           66.0   
77  E0026    Zara        Informatika  24.0         90.4           82.0   

   tanggal_daftar  
28     2024-11-23  
29     2024-10-03  
30     2024-02-27  
31     2024-01-30  
32     2024-09-05  
33     2024-03-06  
34     2024-11-06  
35     2024-12-31  
36            NaN  
37     2024-06-22  
38     2024

Intinya tampilan head(50) dan tail(50) yang kamu tunjukkan hanyalah preview data setelah proses pembersihan, sehingga kamu bisa memastikan bahwa:

Kolom “umur” sudah berubah menjadi angka yang benar, karena sebelumnya berisi campuran seperti "twenty", "21years", "NaN_string", -5, 120, 150.

Nilai-nilai salah sudah diganti menggunakan mapping yang kamu buat (misalnya "twenty" → 20, "21years" → 21, nilai ekstrem seperti 120, 150, -5, dan "NaN_string" semuanya diganti menjadi 25).

Kolom sudah dalam format yang bersih, misalnya:

umur sudah menjadi float64

tidak ada lagi teks aneh di kolom umur

semua baris terlihat konsisten saat dilihat di head() dan tail()

In [ ]:
Andre_Gerardo_Tarigan['umur'].mean() #mencari rata-rata

np.float64(25.19178082191781)

Ketika kolom umur sudah dinormalisasi (nilai aneh seperti “twenty”, “21years”, “NaN_string”, 150, 120, −5 diganti sesuai mapping lalu diubah menjadi angka), dan nilai yang tetap NaN dibiarkan, fungsi mean() kemudian menghitung rata-rata hanya dari nilai umur yang valid (angka). Hasil akhirnya berupa tipe data np.float64 dengan nilai sekitar 25.19, yang berarti secara keseluruhan umur mahasiswa dalam dataset berada pada kisaran 25 tahun.

In [ ]:
Andre_Gerardo_Tarigan['umur'] = mhs_df['umur'].fillna(22) #isi dengan nilai rata-rata

In [ ]:
Andre_Gerardo_Tarigan.isna().sum()

nim               0
nama              0
jurusan           0
umur              5
nilai_akhir       5
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Intinya, masalahnya terjadi karena kamu mengisi nilai NaN pada kolom **umur** menggunakan dataframe **mhs_df**, tetapi yang kamu cek dan gunakan sebenarnya adalah dataframe **Andre_Gerardo_Tarigan**. Karena dua dataframe itu berbeda, perintah:

```python
mhs_df['umur'].fillna(22)
```

tidak mengubah apa pun di dataframe **Andre_Gerardo_Tarigan**, sehingga ketika kamu memeriksa lagi NaN-nya, kolom **umur** tetap menunjukkan masih ada **5 nilai kosong**. Solusinya cukup dengan mengisi NaN langsung pada dataframe yang benar, yaitu:

```python
Andre_Gerardo_Tarigan['umur'] = Andre_Gerardo_Tarigan['umur'].fillna(22)
```

Dengan begitu semua NaN akan terisi, dan pengecekan ulang akan menghasilkan nilai NaN = 0.


In [ ]:
Andre_Gerardo_Tarigan.duplicated().sum()

np.int64(8)

Kode **Andre_Gerardo_Tarigan.duplicated().sum()** menunjukkan bahwa di dalam dataset terdapat **8 baris yang duplikat**, artinya ada 8 baris data yang isinya sama persis dengan baris lain. Ini penting karena data duplikat bisa membuat analisis menjadi bias, misalnya nilai rata-rata akan terpengaruh karena data yang sama dihitung lebih dari sekali. Dengan mengetahui jumlah duplikasi, kita bisa memutuskan apakah baris-baris tersebut perlu dihapus agar dataset lebih bersih dan akurat.


In [ ]:
Andre_Gerardo_Tarigan.head(55)


nim    nama            jurusan   umur  nilai_akhir  kehadiran_pct  \
0   E0001    Aldi             Teknik  150.0         59.5           82.0   
1   E0002    Budi             teknik   20.0         84.5           99.0   
2   E0003   Citra             Energi   -5.0         56.2           78.0   
3   E0004    Dewi             Teknik    NaN         63.0           99.0   
4   E0005     Eka   Manajemen Energi   22.0        999.0           65.0   
5   E0006   Fajar             Energi   21.0        -20.0           78.0   
6   E0007    Gina        Informatika   19.0         70.2          150.0   
7   E0008  Hendra             Energi   25.0         56.7          -10.0   
8   E0009     Ika        Informatika  120.0         58.2           83.0   
9   E0010    Joko             Teknik   23.0         62.7           69.0   
10  E0011    Kiki             Energi   25.0         66.6           70.0   
11  E0012    Lina        Informatika   22.0         66.1            NaN   
12  E0013    Maya    ManajemenEnergi   21.0         63.4           82.0   
13  E0014   Nando             Teknik   23.0         80.4           99.0   
14  E0015     Oki            energi    22.0         70.2          100.0   
15  E0016    Putu             Energi   25.0         85.7           68.0   
16  E0017    Qori             Teknik   25.0         67.6           82.0   
17  E0018    Rina             EnErGi   23.0         90.1           85.0   
18  E0019    Sari        Informatika   22.0         90.4           94.0   
19  E0020    Tono        Informatika   20.0         64.9          100.0   
20  E0021     Uli             Teknik   24.0         78.4           90.0   
21  E0022    Vina             Energi    NaN         75.4           70.0   
22  E0023   Wawan             Energi   20.0         80.1           92.0   
23  E0024    Xena        Informatika   22.0         70.3           94.0   
24  E0025    Yudi  Pembangkit Energi   23.0         89.4           65.0   
25  E0026    Zara        Informatika   24.0         90.4           82.0   
26  E0027    Arif             Energi   19.0         66.7           75.0   
27  E0028    Bima             Teknik   24.0         90.0           81.0   
28  E0029    Cici             Teknik   23.0         91.5           84.0   
29  E0030    Dani             Energi   20.0          NaN           75.0   
30  E0031    Eros        Informatika   23.0         68.0           66.0   
31  E0032   Ferry             Teknik   25.0         87.3           80.0   
32  E0033  Gilang             Energi   23.0          NaN           70.0   
33  E0034    Hani   Manajemen Energi   25.0         76.3           73.0   
34  E0035   Indra   Manajemen Energi   23.0         61.6           98.0   
35  E0036    Jeni        Informatika   20.0         92.2           77.0   
36  E0037    Kris   Manajemen Energi   22.0         94.8           88.0   
37  E0038    Lulu             Energi   19.0         64.9           69.0   
38  E0039    Mita             Energi   23.0         77.2           79.0   
39  E0040    Nina             Energi   24.0         80.2           79.0   
40  E0041    Omar        Informatika   25.0          NaN           97.0   
41  E0042     Pia   Manajemen Energi   24.0         74.4           99.0   
42  E0043    Rico             Energi   20.0         86.4           91.0   
43  E0044   Sinta   Manajemen Energi   19.0         81.4           87.0   
44  E0045    Tari             Energi   22.0         89.6           68.0   
45  E0046    Umar             Teknik   19.0         71.1           86.0   
46  E0047    Vita   Manajemen Energi   20.0         62.6           93.0   
47  E0048    Widi   Manajemen Energi   20.0         66.1           80.0   
48  E0049     Xul             Energi   25.0         89.5          100.0   
49  E0050   Yosef             Teknik   24.0         95.0           65.0   
50  E0051    Zaki   Manajemen Energi   20.0         71.3           95.0   
51  E0052    Adel        Informatika    NaN         61.6           65.0   
52  E0053   Bagus   Manajemen E



Output `Andre_Gerardo_Tarigan.duplicated().sum()` yang menghasilkan nilai **8** berarti bahwa di dalam DataFrame-mu terdapat **8 baris yang terdeteksi sebagai duplikat** secara keseluruhan. Fungsi `.duplicated()` mengecek setiap baris dan membandingkannya dengan baris sebelumnya; jika semua kolom pada suatu baris sama persis dengan baris lain yang sudah muncul sebelumnya, maka baris itu dianggap duplikat. Jadi, angka 8 menunjukkan ada 8 baris yang berulang dari baris-barais lain.

Kemudian, ketika kamu menampilkan `Andre_Gerardo_Tarigan.head(55)`, terlihat bahwa datasetmu memang memiliki berbagai penyimpangan data seperti umur 150, umur -5, nilai_akhir 999 atau -20, kehadiran_pct negatif, tanggal yang tidak konsisten formatnya, hingga entri jurusan yang bercampur-mix case atau salah ketik. Semua anomali ini tidak menyebabkan error pada `.duplicated()`, tetapi bisa menyebabkan duplikasi tidak hanya karena pengulangan data, melainkan juga bisa menyulitkan pembersihan data secara menyeluruh jika tidak ditangani. Namun, dalam konteks duplikasi, pandas hanya mendeteksi baris yang benar-benar identik secara nilai—bukan hanya mirip atau salah ketik.

Kesimpulannya, angka 8 itu berarti ada **8 baris identik** secara penuh di datasetmu, dan screenshot DataFrame menunjukkan kondisi data mentah yang masih perlu dibersihkan lebih jauh, tetapi tidak mempengaruhi makna dari output `.duplicated().sum()` tersebut.


In [ ]:
Andre_Gerardo_Tarigan.tail()

nim   nama      jurusan  umur  nilai_akhir  kehadiran_pct tanggal_daftar
73  E0057  Fauzi       Energi  23.0         58.2           76.0     2024-02-05
74  E0039   Mita       Energi  23.0         77.2           79.0     2024-01-04
75  E0069  Salsa  Informatika  23.0         63.4           86.0     2024-04-14
76  E0031   Eros  Informatika  23.0         68.0           66.0     2024-02-27
77  E0026   Zara  Informatika  24.0         90.4           82.0     2024-11-24

Output yang kamu tampilkan adalah hasil dari **Andre_Gerardo_Tarigan.tail()**, yaitu perintah Pandas untuk menampilkan **5 baris terakhir** dari DataFrame. Artinya, kamu sedang melihat data mahasiswa yang berada di urutan paling akhir setelah proses pembersihan atau pengurutan sebelumnya.

Dalam tampilan tersebut terlihat bahwa:

* **Baris indeks 73–77** adalah data paling akhir dalam DataFrame.
* Setiap baris menampilkan kolom:

  * **nim**
  * **nama**
  * **jurusan**
  * **umur**
  * **nilai_akhir**
  * **kehadiran_pct**
  * **tanggal_daftar**
* Semua data pada lima baris terakhir ini sudah tampak bersih dan tidak ada nilai aneh seperti umur negatif, nilai kehadiran >100%, atau tanggal dalam format salah.
* Ini membantu memastikan bahwa setelah proses pembersihan data, bagian akhir dataset juga berada dalam kondisi yang konsisten dan tidak terdapat data bermasalah.

Jadi secara ringkas, **tail() menampilkan lima baris terakhir DataFrame untuk memastikan data bagian bawah juga valid dan rapi**.


In [ ]:
Andre_Gerardo_Tarigan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             78 non-null     object 
 1   nama            78 non-null     object 
 2   jurusan         78 non-null     object 
 3   umur            73 non-null     float64
 4   nilai_akhir     73 non-null     float64
 5   kehadiran_pct   76 non-null     float64
 6   tanggal_daftar  76 non-null     object 
dtypes: float64(3), object(4)
memory usage: 4.4+ KB




Ketika kamu menampilkan `Andre_Gerardo_Tarigan.isna().sum()`, terlihat bahwa beberapa kolom memiliki nilai *missing* (NaN). Khusus kolom **umur**, terdapat 5 nilai yang kosong. Karena NaN ini dapat mengganggu analisis, kamu kemudian menghitung rata-rata umur dengan `.mean()` dan didapat nilai sekitar **25.19**. Namun pada kode berikutnya, kamu memilih mengisi NaN dengan angka **22**, sehingga ketika dicek lagi jumlah NaN, kolom umur masih menunjukkan angka 5. Ini terjadi karena kamu menulis:

```python
Andre_Gerardo_Tarigan['umur'] = mhs_df['umur'].fillna(22)
```

yang artinya kamu tidak mengisi kolom di DataFrame **Andre_Gerardo_Tarigan**, tetapi mengambil kolom *umur* dari **mhs_df**. Akibatnya kolom umur di `Andre_Gerardo_Tarigan` tetap tidak berubah.

Lalu, ketika kamu mengecek duplikasi menggunakan:

```python
Andre_Gerardo_Tarigan.duplicated().sum()
```

muncul angka **8**, yang berarti ada **8 baris data yang sama persis** di dalam dataset.

Setelah itu, kamu melihat sebagian isi dataset dengan `.head(55)` dan `.tail()`. Dari tampilan awal data, terlihat beberapa *outlier* seperti:

* umur **150**, **120**, **-5**, atau **NaN**
* nilai_akhir ada yang **999** dan **-20**
* kehadiran_pct ada yang **-10**
* jurusan ditulis tidak konsisten (“Teknik”, “teknik”, “EnErGi”, dll)
* tanggal daftar ada yang tidak dalam format tanggal (contoh “05/02/2024”)

Ini menunjukkan dataset masih kotor dan membutuhkan pembersihan lebih lanjut.

Terakhir, ketika kamu menampilkan:

```python
Andre_Gerardo_Tarigan.info()
```

hasilnya menunjukkan:

* total baris = **78**
* ada **3 kolom numerik** (umur, nilai_akhir, kehadiran_pct)
* ada **4 kolom object** (nim, nama, jurusan, tanggal_daftar)
* masih ada **nilai kosong (NaN)**:

  * umur = 73 non-null → berarti masih ada **5 NaN**
  * nilai_akhir = 73 non-null → berarti **5 NaN**
  * kehadiran_pct = 76 non-null → berarti **2 NaN**
  * tanggal_daftar = 76 non-null → berarti **2 NaN**

Artinya dataset belum bersih dan butuh pengolahan tambahan.

---

### **Kesimpulan keseluruhan (dalam satu alur penjelasan)**

Dataset milikmu memiliki 78 data, tetapi masih terdapat berbagai masalah seperti nilai kosong, kesalahan penulisan, format tanggal tidak konsisten, nilai ekstrim seperti umur 150 atau nilai -20, serta adanya 8 baris duplikat. Upaya pengisian NaN yang kamu lakukan belum berhasil karena kamu mengisi kolom dari DataFrame lain. Semua ringkasan ini terlihat jelas dari hasil `.head()`, `.tail()`, hasil duplikasi, serta output `DataFrame.info()` yang menampilkan jumlah non-null di setiap kolom. Dengan kondisi seperti ini, dataset perlu dibersihkan lebih lanjut agar dapat digunakan dengan benar untuk analisis berikutnya.

